include/mfmg/hierarchy.hpp

```diff
@@ -0,0 +1,146 @@
+  }
+  std::shared_ptr<const Operator> get_smoother() const { return _smoother; }
+
+  void set_operator(std::shared_ptr<const Operator> op) { _operator = op; }
+  void set_restrictor(std::shared_ptr<const Operator> r) { _restrictor = r; }
+  void set_prolongator(std::shared_ptr<const Operator> p) { _prolongator = p; }
+  void set_smoother(std::shared_ptr<const Operator> s) { _smoother = s; }
+
+private:
+  std::shared_ptr<const Operator> _operator, _prolongator, _restrictor,
+      _smoother;
+};
+
+template <class MeshEvalutor, class Operator, class Vector, class Device>
+class Hierarchy
+{
+public:
+  using mesh_type = typename MeshEvalutor::mesh_type;
+  using vector_type = Vector;
+  using operator_type = Operator;
+  using device_type = Device;
+
+public:
+  Hierarchy(MPI_Comm comm, MeshEvalutor &evaluator, mesh_type &mesh,
+            std::shared_ptr<boost::property_tree::ptree> params = nullptr)
+      : _comm(comm)
+  {
+    static_assert(
+        std::is_same<typename MeshEvalutor::operator_type, Operator>::value,
+        "Operator template does not match the one in MeshEvaluator");
+    static_assert(
+        std::is_same<typename MeshEvalutor::vector_type, Vector>::value,
+        "Vector template does not match the one in MeshEvaluator");
+
+    const int num_levels = 2;
+    _levels.resize(2);
+
+    _levels[0].set_operator(evaluator.get_operator());
+    for (int i = 0; i < num_levels; i++)
+    {
+      auto level = _levels[i];
+      auto level_op = level.get_operator();
+
+      auto restrictor =
+          Adapter<operator_type, vector_type, mesh_type,
+                  device_type>::build_restrictor(_comm, evaluator, level_op,
+                                                 mesh, params);
+      // auto prolongator = details::transpose(restrictor);
+      // auto op_c = details::build_coarse_matrix(op, prolongator, restrictor);
+      // auto smoother = details::build_smoother(op);
+
+      level.set_restrictor(restrictor);
+      // level.set_prolongator(prolongator);
+      // level.set_operator(op_c);
+      // level.set_smoother(smoother);
+    }
+  }
+
+  void apply(const Vector &b, Vector &x, int start_level = 0) const
+  {
+    auto const num_levels = _levels.size();
+
+    auto level = _levels[start_level];
+
+    if (start_level == num_levels - 1)
+    {
+      // Coarsest level
+
+      auto op = level.get_a();
+      // FIXME: do coarse solve
+    }
+    else
+    {
+      auto level = _levels[start_level];
+      auto level_c = _levels[start_level];
+
+      auto op = level.get_operator();
+      auto op_c = level_c.get_operator();
+      auto prolongator = level_c.get_prolongator();
+      auto restrictor = level_c.get_restrictor();
+
+      // apply pre-smoother
+      auto smoother = level_c.get_smoother();
+      smoother->apply(x, b);
+
+      // compute residual
+      auto r = residual(op, x, b);
+
+      // restrict residual
+      Vector b_coarse;
+      r->apply(r, b_coarse);
```

Comment:
Are you sure about this line? Why isn't `r` a vector?

Reply:
Don't mind that ;) The notation is wrong there, it's both residual and restriction matrix...

Reply:
OK